### PR TITLE
[Groundhog] An EVTM listener implementation + basic testing framework

### PR DESCRIPTION
**Pull request by shubhagrawal30**
Groundhog and BITServer code for receiving UDP multicast through the LOS/biphase and TDRSS/highrate EVTM downlinks. Also, starting a testing framework for groundhog.

Sender in MCP was implemented in #63.  

### Scope:

- _content_ changes made for new groundhog `evtm` code
- no _content_ changes made to the old groundhog code. Overall, several formatting changes to pass the _style checker/linter_
- adding `CMakeList` stuff for testing
- exhaustive (?) test code for the new `evtm` code
- no test code (yet) for the other downlinks
- no changes for possible integration with code for Iridium Pilot

### TODOs for new GSs or FCs for testing:

1. Configure `ufw` firewall on the ground station: **run `sudo ufw allow 20001/udp` and `sudo ufw allow 20002/udp` to get incoming UDP traffice on the relevent ports!** If you also want `pilot` telemetry, do the same but with ports 31213 and 31214.
2. On your flight computer, add a default gateway to the gondola network config in `/etc/netplan/00-installer-config.yaml`: 
&nbsp;  &nbsp;  `routes:`
&nbsp;  &nbsp;  &nbsp;  &nbsp;  `- to: default`
&nbsp;  &nbsp;  &nbsp;  &nbsp;  &nbsp;  `via: 192.168.1.1`

### CODE REVIEW

File: groundhog/blast/main.c [hunk 1]
```diff
@@ -39,56 +48,79 @@ int main(int argc, char * argv[]) {
...
   generate_housekeeping_linklist(linklist_find_by_name(ALL_TELEMETRY_NAME, ll_list), ALL_TELEMETRY_NAME);
-  linklist_generate_lookup(ll_list);  
-  write_linklist_format(linklist_find_by_name(ALL_TELEMETRY_NAME, ll_list), DEFAULT_LINKLIST_DIR ALL_TELEMETRY_NAME ".auto");
+  linklist_generate_lookup(ll_list);
+  write_linklist_format(linklist_find_by_name(ALL_TELEMETRY_NAME, ll_list), \
+                        DEFAULT_LINKLIST_DIR ALL_TELEMETRY_NAME ".auto");
 
   channels_write_calspecs("test.cs", derived_list);
 
   int pilot_on = 1;
```
Comment (ianlowe13) [inline groundhog/blast/main.c:57]: deprecated comment

File: groundhog/blast/main.c [hunk 2]
```diff
@@ -39,56 +48,79 @@ int main(int argc, char * argv[]) {
...
   int bi0_on = 1;
   int highrate_on = 1;
+  int evtm_los_on = 1;
+  int evtm_tdrss_on = 1;
   int daemon = 0;
 
   for (int i = 1; i < argc; i++) {
     if (strcmp(argv[i], "-no_pilot") == 0) pilot_on = 0;
```
Comment (ianlowe13) [inline groundhog/blast/main.c:65]: more comments about deprecated modes to be removed

**Review by ianlowe13 — COMMENTED** [review]
see comments

File: common/bitserver.c [hunk 3]
```diff
@@ -435,18 +435,41 @@ int initBITRecver(struct BITRecver *server, const char *recv_addr,
...
+  if (strcmp(net_addr, MULTICAST_ADDR_START) >= 0 && strcmp(net_addr, MULTICAST_ADDR_END) <= 0) {
+    // set up multicast address
+    struct ip_mreq mreq;
+    mreq.imr_multiaddr.s_addr = inet_addr(recv_addr);
+    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+    if (setsockopt(server->sck, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) < 0) {
+      blast_err("unable to set multicast address: %s:%d", recv_addr, port);
+      return -1;
+    }
+    blast_info("-> Multicast mode: %s:%d", recv_addr, port);
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply common/bitserver.c:472]: not really, changed it.

File: groundhog/blast/highrate.c [hunk 4]
```diff
@@ -1,3 +1,9 @@
+/**
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/highrate.c:1]: yeah, I do not know who wrote this/copyright. This is needed to pass the linter though (groundhog was not style-checked before as it did not have a testing framework)

File: groundhog/blast/highrate.c [hunk 5]
```diff
@@ -258,130 +262,128 @@ void highrate_receive(void *arg) {
...
+            
+            } else {
+                blast_info("[%s] Bad syncword 0x%.08x\n", source_str, *(uint32_t *) gse_packet);
+            }
+            */
+        }
       }
   }
 }
-
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/highrate.c:387]: I don't know how this passed the linter, haha

**Comment by shubhagrawal30** [comment]
Also adding in EVTM commanding code! tangentially related to this PR overall, but still a good addition to review here.

File: groundhog/blast/highrate.c [hunk 6]
```diff
@@ -239,7 +243,7 @@ void highrate_receive(void *arg) {
           // mimic the return header values for highrate
           gse_packet_header.size = 2048;
```
Comment (ianlowe13) [inline groundhog/blast/highrate.c:244]: magic numbers should be #define

**Review by ianlowe13 — CHANGES_REQUESTED** [review]
Shubh, do you know if the pilot link has any "in charge" protections? Maybe in bitsender? I think all links should only send telemetry if InCharge == 1 but it would be nice to double check.

**Comment by shubhagrawal30** [comment]
> Shubh, do you know if the pilot link has any "in charge" protections? Maybe in bitsender? I think all links should only send telemetry if InCharge == 1 but it would be nice to double check.

Yup. In either one's infinite loop, stuff is only compressed and sent to BITSender when the thread is running on the InCharge computer.

`if (!fifoIsEmpty(es->evtm_fifo) && es->ll && InCharge) { // data ready to send` in `evtm.c:164`

`if (!fifoIsEmpty(&pilot_fifo) && ll && InCharge) { // data is ready to be sent` in `pilot.c:116`

**Review by evanmayer — CHANGES_REQUESTED** [review]

File: common/bitserver.c [hunk 7]
```diff
@@ -428,18 +435,41 @@ int initBITRecver(struct BITRecver *server, const char *recv_addr,
...
-    blast_err("unable to set reusable port");
+  if (setsockopt(server->sck, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) == -1) {
+    blast_err("unable to set reusable port: %s:%d", recv_addr, port);
+  }
+
+  // check if the recv_addr IP is a multicast address, if it is then set the socket to multicast
+  // multicast addressed are in the range 224 to 239
+  const char* dot = strchr(recv_addr, '.');
+  char net_addr[4];
+  if (dot != NULL) {
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply common/bitserver.c:456]: The best/easiest thing to do is to just check if the dot address is within 4 of the recv_addr, I have added that.

**Comment by evanmayer** [comment]
For future reference, reviewers should resolve their comments when they're satisfied with your changes (after a commit), or once they agree with your pushback on changes that you argue aren't necessary. This helps make sure that reviewers know their comments have been addressed. But it does require reviewers to be attentive.

File: groundhog/blast/tests/test_gh_evtm.c [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+    assert_null(es.ll);
+    assert_int_equal(es.bad_serial_count, 1);
+    assert_int_equal(es.serial, INCORRECT_SERIAL);
+}
+
+/**
+ * @brief test EVTM_receiver_loop_body when the linklist is not filelist
+ */
+void test_GH_EVTM_receiver_loop_body(void **state) {
+    for (int evtm_type = LOS_EVTM; evtm_type <= TDRSS_EVTM; evtm_type++) {
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/tests/test_gh_evtm.c:287]: would this concern be gone with the use of a few strategically placed printf's?

[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/tests/test_gh_evtm.c:287]: Nvm, I can do this another way

[the remark this answers is not in the document]
Comment (evanmayer) [reply groundhog/blast/tests/test_gh_evtm.c:287]: Can you functionalize this loop body and break each test case out into a separate call to `cmocka_unit_test`? printfs may help track down what failed, or they may get lost in the noise, but the problem here is that if you lump all your tests together (hyperbole) under one `cmocka_unit_test` call, the unit test will report as failed and stop executing if **any** assertion is failed, so you won't have a good idea about which things failed and which things didn't. With separate test invocations for each separate case being tested, even the non-failing tests scheduled after the failing test will still be executed.

**Comment by shubhagrawal30** [comment]
Ah sure, we could do that. I meant to _resolve_ them as a way to track that I have addressed them. But the other way also sounds fine.

**Comment by shubhagrawal30** [comment]
> Can you functionalize this loop body and break each test case out into a separate call to `cmocka_unit_test`? printfs may help track down what failed, or they may get lost in the noise, but the problem here is that if you lump all your tests together (hyperbole) under one `cmocka_unit_test` call, the unit test will report as failed and stop executing if **any** assertion is failed, so you won't have a good idea about which things failed and which things didn't. With separate test invocations for each separate case being tested, even the non-failing tests scheduled after the failing test will still be executed. 

yup, that is what I have done now.

File: mcp/commanding/commands.c [hunk 9]
```diff
@@ -743,6 +743,29 @@ void SingleCommand(enum singleCommand command, int scheduled)
...
+        case disable_evtm_los:
+            CommandData.evtm_los_enabled = 0;
+            break;
+        case enable_evtm_tdrss:
+            CommandData.evtm_tdrss_enabled = 1;
+            break;
+        case disable_evtm_tdrss:
+            CommandData.evtm_tdrss_enabled = 0;
+            break;
+        case enable_evtm:
```
[the remark this answers is not in the document]
Comment (ianlowe13) [reply mcp/commanding/commands.c:760]: same for disable

**Review by ianlowe13 — DISMISSED** [review]

File: groundhog/blast/evtm.c [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            return;
+    }
+
+    // hijacking frame number for transmit size
+    es->transmit_size = es->udprecver.frame_num;
+
+    if (groundhog_check_for_fileblocks(es->ll, FILE_LINKLIST)) {
+        // unpack and extract to disk
+        es->framenum = groundhog_unpack_fileblocks(es->ll, es->transmit_size, es->compbuffer, NULL,
+                                                                                             NULL, NULL, NULL, GROUNDHOG_EXTRACT_TO_DISK);
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/evtm.c:116]: close to function opening "("? at least that is the rough convention I have been going off from. If we don't like that, let me know.

File: groundhog/blast/tests/test_gh_evtm.c [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+    memcpy(*(state+3), &incorrect_serial, sizeof(uint32_t));
+    return 0;
+}
+
+/**
+ * @brief teardown configuration for testing EVTM groundhog code
+ * 
+ * @return status code
+ */
+static int GH_EVTM_teardown_tests(void **state) {
```
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply groundhog/blast/tests/test_gh_evtm.c:182]: I was infact leaking memory there. I could not figure out how to fix it directly though (`free`-ing `state+2` and `state+3` was weirdly seg-faulting). So I did the normal perfectly reasonable thing and rewrote that portion of code.

File: mcp/commanding/commands.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (shubhagrawal30) [reply mcp/commanding/commands.c]: done

**Review by ianlowe13 — COMMENTED** [review]
My issues have been addressed in new changes

**Review by evanmayer — APPROVED** [review]

**Review by ianlowe13 — APPROVED** [review]

**Comment by evanmayer** [comment]
Update: looks like the flight computer had a problem with the edits to the above netplan file, and we had to remove the file from `/etc/netplan` dir to get the star camera computer, which we were using as a ground station, to connect to anything. We tested reception of telemetry in this situation and it seems that the above changes are not necessary...

**Pull request merged**